educe/stac/graph.py

```diff
@@ -197,6 +192,92 @@ def strip_cdus(self, sloppy=False):
         # remove the actual CDU objects too
         self.doc.schemas = [s for s in self.doc.schemas if not stac.is_cdu(s)]
 
+    def strip_cdus_distribute(self, sloppy=False, mode='broadcast'):
+        """ Delete all CDUs in this graph.
+            Links involving a CDU will point to/from all the elements
+            of this CDU.
+            There will be one edge copy for every new source-target
+            combination.
+
+            WARNING: Disabled link-rewriting for annotation layer
+        """
+
+        # Set of labels for which the source node should be distributed
+        LEFT_DIST = frozenset((
+            'Acknowledgement',
+            'Explanation',
+            'Comment',
+            'Continuation',
+            'Narration',
+            'Contrast',
+            'Parallel',
+            'Background'))
+
+        # Set of labels for which the target node should be distributed
+        RIGHT_DIST = frozenset((
+            'Result',
+            'Continuation',
+            'Narration',
+            'Comment',
+            'Contrast',
+            'Parallel',
+            'Background',
+            'Elaboration'))
+
+        # Warning: heads.keys() are hyperedges
+        heads = self.recursive_cdu_heads(sloppy)
+
+        def distrib_candidates(links, attrs):
+            """ Return a pair of list of nodes to be attached,
+                depending on the edge label.
+            """
+            src_node, tgt_node = links
+            label = dict(attrs)['annotation'].type
+
+            def candidates(node, label_set):
+                if self.is_edu(node):
+                    return [node]
+                if label in label_set or mode == 'broadcast':
+                    # Either distribute over all components...
+                    nodes = edu_components(node)
+                else:
+                    # ... or link to the CDU recursive head only
+                    nodes = [heads[self.mirror(node)]]
+                return nodes
+
+            return [candidates(node, lset) for (node, lset) in
+                    ((src_node, LEFT_DIST), (tgt_node, RIGHT_DIST))]
+
```

Comment:
The list comprehension seems superfluous and miscarries the intended return type (a pair).
I would prefer the following:
```
return (candidates(src_node, LEFT_DIST),
            candidates(tgt_node, RIGHT_DIST))
```